influxdb3_server/src/http/v1.rs

```diff
@@ -0,0 +1,485 @@
+    fn flush_one(&mut self) -> Option<(String, Vec<Row>)> {
+        if !self.can_flush() {
+            return None;
+        }
+        // we can flush, so unwrap is safe:
+        let size = self.size.unwrap();
+        if self
+            .series
+            .back()
+            .is_some_and(|(_, rows)| rows.len() <= size)
+        {
+            // the back series is smaller than the chunk size, so we just
+            // pop and take the whole thing:
+            self.series.pop_back()
+        } else {
+            // only drain a chunk's worth from the back series:
+            self.series
+                .back_mut()
+                .map(|(name, rows)| (name.to_owned(), rows.drain(..size).collect()))
+        }
+    }
+
+    /// The [`ChunkBuffer`] is operating in chunked mode, and can flush a chunk
+    fn can_flush(&self) -> bool {
+        if let (Some(size), Some(m)) = (self.size, self.series.back()) {
+            m.1.len() >= size || self.series.len() > 1
+        } else {
+            false
+        }
+    }
+
+    /// The [`ChunkBuffer`] is empty
+    fn is_empty(&self) -> bool {
+        self.series.is_empty()
+    }
+}
+
+/// A wrapper around a [`SendableRecordBatchStream`] that buffers streamed
+/// [`RecordBatch`]es and outputs [`QueryResponse`]s
+///
+/// Can operate in `chunked` mode, in which case, the `chunk_size` provided
+/// will determine how often [`QueryResponse`]s are emitted. When not in
+/// `chunked` mode, the entire input stream of [`RecordBatch`]es will be buffered
+/// into memory before being emitted.
+///
+/// `pretty` will emit pretty formatted JSON.
+///
+/// Providing an `epoch` [`Precision`] will have the `time` column values emitted
+/// as UNIX epoch times with the given precision.
+///
+/// The input stream is wrapped in [`Fuse`], because of the [`Stream`] implementation
+/// below, it is possible that the input stream is polled after completion.
+struct QueryResponseStream {
+    buffer: ChunkBuffer,
+    input: Fuse<SendableRecordBatchStream>,
+    column_map: HashMap<String, usize>,
+    statement_id: usize,
+    pretty: bool,
+    epoch: Option<Precision>,
+}
+
+impl QueryResponseStream {
+    /// Create a new [`QueryResponseStream`]
+    ///
+    /// Specifying a `chunk_size` will have the stream operate in `chunked` mode.
+    fn new(
+        statement_id: usize,
+        input: SendableRecordBatchStream,
+        chunk_size: Option<usize>,
+        pretty: bool,
+        epoch: Option<Precision>,
+    ) -> Result<Self, anyhow::Error> {
+        let buffer = ChunkBuffer::new(chunk_size);
+        let schema = input.schema();
+        let column_map = schema
+            .fields
+            .iter()
+            .map(|f| f.name().to_owned())
+            .enumerate()
+            .flat_map(|(i, n)| {
+                if n != INFLUXQL_MEASUREMENT_COLUMN_NAME && i > 0 {
+                    Some((n, i - 1))
+                } else {
+                    None
+                }
```

Comment:
This assumes the `iox::measurement` column is always first, but things would break if that were not the case. Maybe an error should be thrown if that were the case.